File: NeetCode 150/Backtracking/40. Combination Sum II.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        subset = []
        candidates.sort() # Sort to handle duplicates and pruning

        def backtrack(start,target):
            if target == 0:
                res.append(subset[:])
                return
            for i in range(start,len(candidates)):
                # Prune search if current number exceeds remaining target.
                #All numbers ahead are more than target
                if candidates[i] > target:
                    break
                
                # Skip duplicates at the same recursion level
                # i > start ensures we only skip duplicates in the same depth
                if i>start and candidates[i] == candidates[i-1]:
                    continue
                subset.append(candidates[i])
                backtrack(i+1,target-candidates[i])
                subset.pop()

        backtrack(0,target)
        return res
        

File: NeedCode 150 (JAVA)/Backtracking/40. Combination Sum II.java
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    private List<Integer> subset =  new ArrayList<>();

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        backtrack(0,candidates,target);
        return res;
    }

    private void backtrack(int start, int[] candidates, int target){
        if(target == 0){
            res.add(new ArrayList<>(subset));
        }

        for(int i = start;i<candidates.length;i++){
            if(i > start && candidates[i-1] == candidates[i]) continue;
            if(candidates[i] > target) break;

            subset.add(candidates[i]);
            backtrack(i+1,candidates,target - candidates[i]); // now we are using i + 1 to avoid using same number again. Each number can be used only once
            subset.remove(subset.size() - 1);
        }
    }
}
